### ISSUE TITLE
Invalid field layout element config: Invalid field UID

### ISSUE THREAD

**Issue by romainpoirier** [verbb/formie#894]
**Description**
I can see a lot of errors in the web.log file mentioning a `Invalid field layout element config: Invalid field UID` error.
I can see this seems to be related to Formie: `#48 /vendor/verbb/formie/src/controllers/FormsController.php(47): craft\web\Controller->renderTemplate()`.

I tried to fix this using `./craft utils/fix-element-uids/index`, but without any result: `No elements with duplicate UIDs found`.
I can't confirm if this affect the front-end or back-end: for the moment, it seems a silent error. Why and how to fix this?

```
[-][1][-][warning][craft\models\FieldLayoutTab::init] Invalid field layout element config: Invalid field UID: 9f9ca002-eed5-4270-b20d-29974a8f0d15
[-][1][-][error][yii\base\InvalidArgumentException] yii\base\InvalidArgumentException: Invalid field UID: 9f9ca002-eed5-4270-b20d-29974a8f0d15 in /vendor/craftcms/cms/src/fieldlayoutelements/CustomField.php:109
Stack trace:
#0 /vendor/yiisoft/yii2/base/BaseObject.php(159): craft\fieldlayoutelements\CustomField->setFieldUid()
#1 /vendor/yiisoft/yii2/BaseYii.php(558): yii\base\BaseObject->__set()
#2 /vendor/yiisoft/yii2/base/BaseObject.php(107): yii\BaseYii::configure()
#3 /vendor/craftcms/cms/src/fieldlayoutelements/BaseField.php(59): yii\base\BaseObject->__construct()
#4 /vendor/craftcms/cms/src/fieldlayoutelements/CustomField.php(39): craft\fieldlayoutelements\BaseField->__construct()
#5 [internal function]: craft\fieldlayoutelements\CustomField->__construct()
#6 /vendor/yiisoft/yii2/di/Container.php(419): ReflectionClass->newInstanceArgs()
#7 /vendor/yiisoft/yii2/di/Container.php(170): yii\di\Container->build()
#8 /vendor/yiisoft/yii2/BaseYii.php(365): yii\di\Container->get()
#9 /vendor/craftcms/cms/src/Craft.php(61): yii\BaseYii::createObject()
#10 /vendor/craftcms/cms/src/services/Fields.php(1345): Craft::createObject()
#11 /vendor/craftcms/cms/src/models/FieldLayoutTab.php(141): craft\services\Fields->createLayoutElement()
#12 /vendor/yiisoft/yii2/base/BaseObject.php(109): craft\models\FieldLayoutTab->init()
#13 /vendor/craftcms/cms/src/services/Fields.php(1230): yii\base\BaseObject->__construct()
#14 /vendor/craftcms/cms/src/services/Fields.php(1213): craft\services\Fields->_createLayoutTabFromRow()
#15 [internal function]: craft\services\Fields->craft\services\{closure}()
#16 /vendor/craftcms/cms/src/services/Fields.php(1212): array_map()
#17 /vendor/craftcms/cms/src/models/FieldLayout.php(252): craft\services\Fields->getLayoutTabsById()
#18 /vendor/craftcms/cms/src/models/FieldLayout.php(486): craft\models\FieldLayout->getTabs()
#19 /vendor/craftcms/cms/src/services/ElementIndexes.php(332): craft\models\FieldLayout->getCustomFieldElements()
#20 /vendor/twig/twig/src/Extension/CoreExtension.php(1566): craft\services\ElementIndexes->getSourceSortOptions()
#21 /vendor/craftcms/cms/src/helpers/Template.php(106): twig_get_attribute()
#22 /storage/runtime/compiled_templates/32/32e3afd784b4c197b7f28dc8929dd729048ddbcf65bcc5519b6c50ff9d58b95d.php(160): craft\helpers\Template::attribute()
#23 /vendor/twig/twig/src/Extension/CoreExtension.php(1169): __TwigTemplate_eb2358234cffb03c7b5c645349943dff3a4c1bf2e71673a65e23335427812ef9->macro_sourceLink()
#24 /storage/runtime/compiled_templates/32/32e3afd784b4c197b7f28dc8929dd729048ddbcf65bcc5519b6c50ff9d58b95d.php(89): twig_call_macro()
#25 /vendor/twig/twig/src/Template.php(405): __TwigTemplate_eb2358234cffb03c7b5c645349943dff3a4c1bf2e71673a65e23335427812ef9->doDisplay()
#26 /vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling()
#27 /storage/runtime/compiled_templates/fd/fd2f990976a3f42f49b84cebb26d3b09a4cafc70d7a70d80a77c19d708d826f8.php(106): Twig\Template->display()
#28 /vendor/twig/twig/src/Template.php(182): __TwigTemplate_4ef478ace501e869df5775a3b3b57e7173939463822371097da47d7195617d21->block_sidebar()
#29 /vendor/twig/twig/src/Template.php(254): Twig\Template->displayBlock()
#30 /storage/runtime/compiled_templates/45/453a17d87387399527f58d01dbd560e00d79c4d888402d8cdb9f420dead3f230.php(95): Twig\Template->renderBlock()
#31 /vendor/twig/twig/src/Template.php(405): __TwigTemplate_60a8b492b13bcd9442645b660613839cfe09abdacd3d3e94489948b93abc5b41->doDisplay()
#32 /vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling()
#33 /storage/runtime/compiled_templates/fd/fd2f990976a3f42f49b84cebb26d3b09a4cafc70d7a70d80a77c19d708d826f8.php(71): Twig\Template->display()
#34 /vendor/twig/twig/src/Template.php(405): __TwigTemplate_4ef478ace501e869df5775a3b3b57e7173939463822371097da47d7195617d21->doDisplay()
#35 /vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling()
#36 /storage/runtime/compiled_templates/db/db8d51af9110876b357a010a7d173a2c745f6092fc33f6f8850897de6d1103ab.php(45): Twig\Template->display()
#37 /vendor/twig/twig/src/Template.php(405): __TwigTemplate_924b9289fe05ddbda1ccc389a1cf587c448235f3c9d2bce6aad0ad224e8664e9->doDisplay()
#38 /vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling()
#39 /storage/runtime/compiled_templates/d6/d6a9ffdab1328530ffca5314e8baec8e5b70c7f33789c7c1a290291cf561d314.php(55): Twig\Template->display()
#40 /vendor/twig/twig/src/Template.php(405): __TwigTemplate_a83f44306d2c3857a1cde8772b8b6b0f44463ed572b278728f25b43e67ab0a26->doDisplay()
#41 /vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling()
#42 /vendor/twig/twig/src/Template.php(390): Twig\Template->display()
#43 /vendor/twig/twig/src/TemplateWrapper.php(45): Twig\Template->render()
#44 /vendor/twig/twig/src/Environment.php(318): Twig\TemplateWrapper->render()
#45 /vendor/craftcms/cms/src/web/View.php(408): Twig\Environment->render()
#46 /vendor/craftcms/cms/src/web/View.php(461): craft\web\View->renderTemplate()
#47 /vendor/craftcms/cms/src/web/Controller.php(201): craft\web\View->renderPageTemplate()
#48 /vendor/verbb/formie/src/controllers/FormsController.php(47): craft\web\Controller->renderTemplate()
#49 [internal function]: verbb\formie\controllers\FormsController->actionIndex()
#50 /vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array()
#51 /vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams()
#52 /vendor/yiisoft/yii2/base/Module.php(552): yii\base\Controller->runAction()
#53 /vendor/craftcms/cms/src/web/Application.php(293): yii\base\Module->runAction()
#54 /vendor/yiisoft/yii2/web/Application.php(103): craft\web\Application->runAction()
#55 /vendor/craftcms/cms/src/web/Application.php(278): yii\web\Application->handleRequest()
#56 /vendor/yiisoft/yii2/base/Application.php(384): craft\web\Application->handleRequest()
#57 /htdocs/index.php(26): yii\base\Application->run()
#58 {main}
```

### Form settings

- Multi-page form: Yes
- Submission Method: Page Reload
- Client-side Validation: Yes
- Custom Form Templates: No

**Additional info**
- Plugin version: 1.5.16
- Craft version: 3.7.43
- Multi-site: Yes

**Comment by romainpoirier** [comment]
This is what the Craft CMS support suggested:

> Try that:
> Search the elements column of your fieldlayouttabs for 9f9ca002-eed5-4270-b20d-29974a8f0d15. This will return a row that is a tab in one of your form field layout.
> In your CP, find the form, resave it and the error should go.

However, in the `craft_fieldlayouts` result I can only see an `id` column, but no direct reference to any form + tab. How to reverse this, and find the form to resave?

**Comment by engram-design** [comment]
I believe these are `fieldlayoutelements` leftover from a deleted form, that have failed to be cleaned up. If you search in your `fieldlayouttabs` database table (using `contains` in the `elements` column) for `9f9ca002-eed5-4270-b20d-29974a8f0d15` you'll find the field layout that this field (that no longer exists) belongs to.

They should be deleted when the field layout is deleted, and the `utils/fix-element-uids/index` probably won't do that.

So for example, I've found one on my install for a field with UID `aff842dd-7993-42c6-a5d1-1dbfbfa37378`:

```
4547, 621, Page 1, [{"type":"craft\\fieldlayoutelements\\CustomField","label":null,"instructions":null,"tip":null,"warning":null,"required":false,"width":100,"fieldUid":"aff842dd-7993-42c6-a5d1-1dbfbfa37378"}], 0, 2021-12-18 01:56:41, 2021-12-18 01:56:41, 01e32e7a-7915-4e26-b5e3-1e8c643eb17c
```

But a form with a `fieldLayoutId` of `621` no longer exists, and has been deleted.

**Comment by romainpoirier** [comment]
Yes I can retrieve this field layout in the `fieldlayouttabs` table.
How do you get the `fieldLayoutId` (`621` in your example)? I don't see this column, I only have `layoutId`.
If I search for the `layoutId` value in the `craft_formie_forms` table > `id` column, I don't get any result.

**Comment by engram-design** [comment]
Correct, the `layoutId` is the field layout ID. There's a column called `fieldLayoutId` in `formie_forms` which should match that, and it sounds like you're in the same board, that this fieldlayoutelement belongs to a form that no longer exists.

I use Sequel Pro, but any database UI should be able to search "contains" in the `elements` column.

<img width="950" alt="image" src="https://user-images.githubusercontent.com/1221575/169800049-9f29ee01-6107-4688-8c55-29b50e5e33a4.png">

But you're after the `layoutId` column in the `fieldlayouttabs` table.

Looking up the ID (621 in my example) in `fieldlayouts` you could safely delete that. Formie _should_ be doing that automatically when a form is hard-deleted (for soft-deleted forms, it sticks around so it can be restored), but for some reason these have stuck around.

**Comment by romainpoirier** [comment]
1. `SELECT * FROM `craft_fieldlayouttabs` WHERE `elements` LIKE '%9f9ca002-eed5-4270-b20d-29974a8f0d15%'`
2. Copy the `layoutId` value of the single result
3. Search this value in the `id` column of the `craft_formie_forms` table
4. Got no result

So you can confirm that I can delete safely the row in the `craft_fieldlayouts` got from step `1.`?

**Comment by engram-design** [comment]
That's correct, it'll also cascade to delete the `fieldlayoutelements`

**Comment by romainpoirier** [comment]
Ok, thank you. After cleaning up the web.log file this morning, I can see that I have still more than 200 matches of this error. After checking the Invalid field UID value, I can say that I get a lot of different values. That means that these errors happens a lot of time, and the manual clean up will be very long... Is there any possibility to automating this clean up process? And why are them so much?

**Comment by engram-design** [comment]
No way to automate it (yet), but looking into the why!

**Comment by romainpoirier** [comment]
To be more specify, I can also see that while I have a lot of error like `[error][yii\base\InvalidArgumentException] yii\base\InvalidArgumentException: Invalid field UID: 05aed3df-c43f-4f76-8b0f-29775932aa97` in the `web.log`, most of the time I don't find any result of that in the `craft_fieldlayouttabs` table. So why do I still get these?

**Comment by engram-design** [comment]
I've just added a new console command for you to run. `./craft formie/fields/cleanup-field-layouts` will delete any orphaned field layouts for forms that no longer exist.

Fixed for the next release. To get this fix early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-3 as 1.5.17",
  "...": "..."
}
```

Then run `composer update`. 

I'll look into why this is happening further.

**Comment by romainpoirier** [comment]
For your information, I had to revert quickly in production today from `dev-craft-3 as 1.5.17` to the previous installed version (`1.5.16`) because of two strange behavior:

- On a multi-pages form, notifications were sent after each submitting each step instead of the final submit, while the notification setup was without any condition;
- A notification of a submission last edited a long time ago, and that have been kept untouched since its submission, has been sent without any reasons (while we have tons of others submissions);

Maybe it's not related, but I haven't identified anything else that would cause that since my yesterday update.

Do you see anything since the `1.5.16` version that could be related? Could it be linked to the call of `./craft formie/fields/cleanup-field-layouts`?

**Comment by engram-design** [comment]
So that'd be related to https://github.com/verbb/formie/issues/871 and setting:

```php
Event::on(Submissions::class, Submissions::EVENT_AFTER_INCOMPLETE_SUBMISSION, function(SubmissionEvent $event) {
    $event->handled = false;
});
```

I didn't really think that one through properly, as you say, that'll go on to trigger:

```php
// Fire an 'afterSubmission' event
$event = new SubmissionEvent([
    'submission' => $submission,
    'success' => $success,
]);
$this->trigger(self::EVENT_AFTER_SUBMISSION, $event);

if ($event->success) {
    // Send off some emails, if all good!
    $this->sendNotifications($event->submission);

    // Trigger any integrations
    $this->triggerIntegrations($event->submission);
} else if ($submission->isSpam && $settings->spamEmailNotifications) {
    // Special-case for wanting to send emails for spam
    $this->sendNotifications($event->submission);
}
```

So I'd recommend removing that event until I've come up with a proper way to address that (likely for Formie v2 as it might be a breaking change).

**Comment by romainpoirier** [comment]
What is strange is that:

> On a multi-pages form, notifications were sent after each submitting each step instead of the final submit, while the notification setup was without any condition;

On this form, I have two other notifications' setup, but these one are with conditions (which are based on the status), and this one were not sent at each step of the form, only at the final submit. Do their behavior is different because of their condition? Looks like the condition fix this. I don't understand why the other one is sent at each step on an _incomplete_ form.

> So I'd recommend removing that event until I've come up with a proper way to address that (likely for Formie v2 as it might be a breaking change).

I only have the following, which is a useful dirty fix for undiscovered cases where the `EVENT_AFTER_SUBMISSION` were probably not triggered as expected by default (because of unchanged status at the final submit + notifications not sent + webhooks not triggered).

```
<?php

Event::on(Submission::class, Submission::EVENT_AFTER_SAVE, function(Event $event) {
	$submission = $event->sender;
	$form = $submission->getForm();
	if (!$submission->isIncomplete && $submission->statusId == getenv('FORMIE_STATUS_ON_GOING') && !Craft::$app->getRequest()->getIsCpRequest() && $form->isLastPage()) {
		$notifications = Formie::$plugin->getNotifications()->getFormNotifications($form);
		foreach ($notifications as $notification) {
			$notification = Formie::$plugin->getNotifications()->getNotificationById($notification->id);
			$response = Formie::$plugin->getEmails()->sendEmail($notification, $submission);
		}
	}
});
```

**Comment by engram-design** [comment]
Yep, conditions (if not met) will prevent the notification from sending. But they'll be sent every time you submit the form, if you have that pass-thru (`$event->handled = false;`) enabled, as shown above with that PHP logic. I can recall I advised you to add that, so just ensure that you've removed that.

I don't see an issue with that event logic you've posted. `Submission::EVENT_AFTER_SAVE` will happen every time you save the submission, but you've got appropriate guards in there.

**Comment by romainpoirier** [comment]
Ok, but this is weird to have to add a condition (for example, to match something like the final status of the submission), just to prevent sending unwanted notifications between each submit of a multi-pages form.

I still have this in my module, to prevent unexpected actions while `isIncomplete`:
```
Event::on(Submissions::class, Submissions::EVENT_AFTER_INCOMPLETE_SUBMISSION, function(SubmissionEvent $event) {
	$event->handled = false;
});
```

**Comment by engram-design** [comment]
Yep, I’m saying you’ll want to remove that as I need to redo the logic where that’s called, as it’ll continue on to fire integrations and email notifications when it shouldn’t.

**Comment by romainpoirier** [comment]
Sorry if my explanations aren't very clear. I'm still not sure of what I should do.

> So I'd recommend removing that event until I've come up with a proper way to address that (likely for Formie v2 as it might be a breaking change).

The `$event->handled = false` on `EVENT_AFTER_INCOMPLETE_SUBMISSION` is required in my case to fix unexpected behaviors (notifications not sent / webhooks not triggered). While on `EVENT_AFTER_SAVE` I'm manually sending / triggering them because they are not. But it doesn't seem the issue here because they're not triggered at the time I'm pointing here: they're sent on each step of the multi-pages form.

What should I remove without breaking what's fixed?

**Comment by engram-design** [comment]
Sorry also If I'm not explaining things clearly! My point is that my change to `EVENT_AFTER_INCOMPLETE_SUBMISSION` to fix your other issue was actually flawed, and is the reason why you're getting email notifications and integrations triggered on every step. If you use `$event->handled = false` it'll trigger email notifications and integrations.

However, this should be fixed for the next release. To get this fix early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-3 as 1.5.18",
  "...": "..."
}
```

Then run `composer update`.

**Comment by engram-design** [comment]
Fixed in [1.5.19](https://github.com/verbb/formie/releases/tag/1.5.19)
